### PR TITLE
FlagBagType support

### PR DESCRIPTION
**Pull request by michnovka**
This is initial commit for `FlagBagType` support.

Lets discuss in here

### CODE REVIEW

**Comment by michnovka** [comment]
I have ported also `BitmaskToBitFlagsValueTransformer`, but I dont think this is going to be used now. In fact the whole `as_value` part makes little to no sense for me under the new implementation. One of the issues is that the new 2.x `EnumType` extends the Symfony's own `EnumType` and not the `ChoiceType` which was the case with 1.x version. So if we use `DataTransformer` that spits out `int` and not `BackedEnum[]`, then the Symfony `EnumType` does not handle it and fails.

I have made `BitmaskToBitFlagsValueTransformerTest` so I think we can keep the DataTransformer in the package, maybe somebody will find it useful. But Id remove the `as_value` option from the `FlagBagType` and would only allow it to work with the `FlagBagToCollectionTransformer` (refactored version of `SingleToCollectionFlagEnumTransformer`)

WDYT @ogizanagi ?

**Comment by ogizanagi** [comment]
> I have ported also BitmaskToBitFlagsValueTransformer, but I dont think this is going to be used now. In fact the whole as_value part makes little to no sense for me under the new implementation.

I agree, let's remove this part until it is proven useful. 
I don't want to strictly port every 1.x features as is, only what makes sense regarding the new PHP 8.1 native enums implementation. Also we don't need everything to land in 2.0.0, just what can impact the design and main interfaces.

**Comment by michnovka** [comment]
> Also we don't need everything to land in 2.0.0, just what can impact the design and main interfaces.

Yes, but I would like to include SQL enum, this PR and also FlagBag DBAL type for sure

**Comment by michnovka** [comment]
This is ready for merge @ogizanagi

**Comment by michnovka** [comment]
One important note: I have made the `FlagBag::getBitmask()` public. I think its a useful tool to be used with FlagBags and dont see any reason why it should stay protected.

**Review by benji07 — APPROVED** [review]

File: src/Bridge/Symfony/Form/DataTransformer/BitmaskToBitFlagsValueTransformer.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ * @author Elao <contact@elao.com>
+ */
+
+namespace Elao\Enum\Bridge\Symfony\Form\DataTransformer;
+
+use Elao\Enum\FlagBag;
+use Symfony\Component\Form\Exception\TransformationFailedException;
+
+class BitmaskToBitFlagsValueTransformer extends AbstractFlagBagTransformer
```
Comment (ogizanagi) [inline src/Bridge/Symfony/Form/DataTransformer/BitmaskToBitFlagsValueTransformer.php]: Let's remove it completely

File: src/FlagBag.php [hunk 2]
```diff
@@ -57,6 +51,24 @@ public function __construct(string $enumType, int ...$bits)
         $this->bits = static::decodeBits($bits);
     }
 
+    /**
+     * @param class-string $enumType
+     *
+     * @throws InvalidArgumentException
```
Comment (ogizanagi) [inline src/FlagBag.php]: That's a logic exception, let's not document it

File: src/FlagBag.php [hunk 3]
```diff
@@ -57,6 +51,24 @@ public function __construct(string $enumType, int ...$bits)
...
+     *
+     * @return void
```
Comment (ogizanagi) [inline src/FlagBag.php]: Let's use `void` return type on the signature directly

File: src/FlagBag.php [hunk 4]
```diff
@@ -57,6 +51,24 @@ public function __construct(string $enumType, int ...$bits)
...
+     */
+    public static function checkIntBackedEnumType(string $enumType)
```
Comment (ogizanagi) [inline src/FlagBag.php]: Let's keep this private for now

```suggestion
    private static function checkIntBackedEnumType(string $enumType): void
```

File: src/Bridge/Symfony/Form/Type/FlagBagType.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            ->setDefault('multiple', true)
+            ->setAllowedValues('class', static function (string $enumType): bool {
+                if (!is_a($enumType, \BackedEnum::class, true)) {
+                    return false;
+                }
+
+                try {
+                    // this throws if some value is not power of 2
+                    FlagBag::getBitmask($enumType);
+                } catch (LogicException $e) {
```
Comment (ogizanagi) [inline src/Bridge/Symfony/Form/Type/FlagBagType.php]: I think we should not even care here, the exception will be thrown with a clear enough message for the user to fix it. Let's remove this `setAllowedValues`.

File: tests/Unit/Bridge/Symfony/Form/DataTransformer/FlagBagToCollectionTransformerTest.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        $this->expectExceptionMessage('Expected FlagBag instance of "Elao\Enum\Tests\Fixtures\Enum\Permissions" values. Got FlagBag instance of "Elao\Enum\Tests\Fixtures\Enum\PermissionsMissingBit" values.');
+
+        $transformer = new FlagBagToCollectionTransformer(Permissions::class);
+
+        $transformer->transform(new FlagBag(PermissionsMissingBit::class, FlagBag::NONE));
+    }
+
+    public function provideCollectionToFlagBag(): \Traversable
+    {
+        yield from array_map('array_reverse', iterator_to_array($this->provideFlagBagToCollection()));
```
Comment (ogizanagi) [inline tests/Unit/Bridge/Symfony/Form/DataTransformer/FlagBagToCollectionTransformerTest.php]:
```suggestion
        yield from array_map(array_reverse(...), iterator_to_array($this->provideFlagBagToCollection()));
```

**Comment by michnovka** [comment]
@ogizanagi changes committed, ready for merge

**Comment by ogizanagi** [comment]
Thanks @michnovka for working on this feature, this is much appreciated.

**Comment by michnovka** [comment]
@ogizanagi np, I use this lib so I like to contribute. I will see if I can work on FlagBag DBAL type when https://github.com/Elao/PhpEnums/pull/182 is merged

**Pull request merged**